features/device_key/source/DeviceKey.cpp

```diff
@@ -109,50 +118,57 @@ int DeviceKey::write_key_to_nvstore(uint32_t *input, size_t isize)
     //First we read if key exist. If it is exists, we return DEVICEKEY_ALREADY_EXIST error
     uint32_t read_key[DEVICE_KEY_32BYTE / sizeof(uint32_t)] = {0};
     size_t read_size = DEVICE_KEY_32BYTE;
-    int ret = read_key_from_nvstore(read_key, read_size);
+    int ret = read_key_from_kvstore(read_key, read_size);
     if (DEVICEKEY_SUCCESS == ret) {
         return DEVICEKEY_ALREADY_EXIST;
     }
     if (DEVICEKEY_NOT_FOUND != ret) {
         return ret;
     }
 
-    NVStore &nvstore = NVStore::get_instance();
-    ret = nvstore.set(NVSTORE_DEVICEKEY_KEY, (uint16_t)isize, input);
-    if (NVSTORE_WRITE_ERROR == ret || NVSTORE_BUFF_TOO_SMALL == ret) {
+    KVMap& kv_map = KVMap::get_instance();
+    KVStore * inner_store = kv_map.get_internal_kv_instance(NULL);
+    if (inner_store == NULL) {
         return DEVICEKEY_SAVE_FAILED;
     }
 
-    if (NVSTORE_SUCCESS != ret) {
-        return DEVICEKEY_NVSTORE_UNPREDICTED_ERROR;
+    ret = ((TDBStore*)inner_store)->reserved_data_set(input, isize);
+    if (MBED_ERROR_WRITE_FAILED == ret) {
+        return DEVICEKEY_SAVE_FAILED;
+    }
+
+    if (MBED_SUCCESS != ret) {
+        return DEVICEKEY_KVSTORE_UNPREDICTED_ERROR;
     }
 
     return DEVICEKEY_SUCCESS;
 }
 
-int DeviceKey::read_key_from_nvstore(uint32_t *output, size_t &size)
+int DeviceKey::read_key_from_kvstore(uint32_t *output, size_t& size)
 {
     if (size > (uint16_t) -1) {
         return DEVICEKEY_INVALID_PARAM;
     }
 
-    uint16_t in_size = size;
-    uint16_t out_size = 0;
-    NVStore &nvstore = NVStore::get_instance();
-    int nvStatus = nvstore.get(NVSTORE_DEVICEKEY_KEY, in_size, output, out_size);
-    if (NVSTORE_NOT_FOUND == nvStatus) {
+    KVMap& kv_map = KVMap::get_instance();
+    KVStore * inner_store = kv_map.get_internal_kv_instance(NULL);
+    if (inner_store == NULL) {
+        return DEVICEKEY_NOT_FOUND;
+    }
+
+    int kvStatus = ((TDBStore*)inner_store)->reserved_data_get(output, size);
```

Comment:
Sorry, but this breaks the KVStore abstraction. We should move "reserved_data_get" to the KVStore API. Or should the device key be a KV pair as it was in NVStore?

Reply:
@geky you are completely right. The implementation of root of trust (devicekey) needs to be moved to the PSA area, and all the implementation in terms of API and components needs to be reviewed.
This was not done in 5.11 and will be done in 5.12.
For the moment we have the minimal implementation that can keep backward compatibility of the binary data in storage, meaning we do not want the bootloader to have all TDBStore functionality, but read from a predefined address.
I recommend to go with this implementation and review it later in a wider context.

Reply:
We should still move "reserved_data_get" into the KVStore API, even if it's temporary. (short term decisions have a tendency to stick around).

Reply:
After reviewing this extensively, I think the best solution is the current implementation of reserving ~150 bytes for the ROT instead of reserving a full sector, that in some boards can be very big and in most of the cases will be at least 4k.
Also the reading mechanism for the bootloader will be minimal, instead of using full TDBStore